File: lintcode/easy/nth_to_last_node_in_list/cpp/nth_to_last_node_in_list.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/nth-to-last-node-in-list
@Language: C++
@Datetime: 17-02-15 23:26
*/

/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @param n: An integer.
     * @return: Nth to last node of a singly linked list. 
     */
    ListNode *nthToLast(ListNode *head, int n) 
    {
        if (head == nullptr) {
            return nullptr;
        }
        
        ListNode *slow{ head }, *fast{ head };
        
        while (n-- > 0)
        {
            if (fast->next == nullptr) {
                return slow;
            }
            fast = fast->next;
        }
        
        while (fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next;
        }
        
        return slow->next;
    }
};

File: lintcode/easy/nth_to_last_node_in_list/py/nth_to_last_node_in_list.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/nth-to-last-node-in-list
@Language: Python
@Datetime: 17-02-15 23:23
'''

"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of linked list.
    @param n: An integer.
    @return: Nth to last node of a singly linked list. 
    """
    def nthToLast(self, head, n):
        if head == None:
            return None
            
        slow = head
        fast = head
        
        while n > 0:
            if fast.next == None:
                return slow
                
            fast = fast.next
            n   -= 1
            
        while fast.next != None:
            slow = slow.next
            fast = fast.next
            
        return slow.next